### ISSUE TITLE
Leave import statements unchanged else ensure that re-normalized import paths ends with .js extension, instead of .ts in typescript projects

### ISSUE THREAD

**Issue by harrison-ifeanyichukwu** [rollup/rollup#2745]
First I want to say big thank you to the whole team for this wonderful project.

I am working on a project, setup with **typescript**, **jest**, **rollup**, and **rollup-all**. I am using `babel-preset-typescript` . My target is to build to node.js **cjs** format.

The problem is that after the build, with rollup having replaced **import statements** with node's **require statement**, it adds `.ts` file extension to the imported paths, instead of using import paths as it is, or making sure that the path extension ends with `.js`.

**short code illustration, all files are inside src directory**

- modules/is-string.ts

  ```typescript
  export default (arg: any): arg is string => {
      return typeof arg === 'string';
  }
  ```
- index.ts

  ```typescript
  export {default as isString} from './modules/is-string';
  ```

**Build output (lib/index.js)**

```javascript
  'use strict';

  Object.defineProperty(exports, '__esModule', { value: true });

  var isString_ts = require('./modules/is-string.ts');
 //                                                  see it here  ^^


  exports.isString = isString_ts;
  //# sourceMappingURL=index.js.map
```

I took some time to look at **rollup internals**, and this is happening because **rollup** uses the absolute path returned by **rollup-plugin-node-resolve**, then renormalizes same inside [Chunk.render method](https://github.com/rollup/rollup/blob/master/src/Chunk.ts#L578) without ensuring that path extension ends with `.js`.

Infact, there is no need for rollup to touch relative import paths.

- Rollup Version: **0.66.6**
- Operating System (or Browser): **Ubuntu 18.04**
- Node Version: **10**

### How Do We Reproduce?

- Please clone this [repository](https://github.com/harrison-ifeanyichukwu/utils)

  ```bash
  git clone https://github.com/harrison-ifeanyichukwu/utils.git
  ```
- install dependencies and run build

  ```bash
  npm install && npm run build
  ```

### Expected Behavior

Expected  behaviour is for rollup to just leave the import path as it was in the source file, or it ensures that the path extension ends with **.js** if it decides to tamper with it.

**Expected build output**

```javascript
  'use strict';

  Object.defineProperty(exports, '__esModule', { value: true });

  var isString_ts = require('./modules/is-string');
 //                keep original ^^^^^^^^^^^^^^^^^^


  exports.isString = isString_ts;
  //# sourceMappingURL=index.js.map
```

**OR**

```javascript
  'use strict';

  Object.defineProperty(exports, '__esModule', { value: true });

  var isString_ts = require('./modules/is-string.js');
 // ensure it ends with .js    ^^^^^^^^^^^^^^^^^^^^


  exports.isString = isString_ts;
  //# sourceMappingURL=index.js.map
```
### Actual Behavior

rollup adds `.ts` to imported path, something it picks from **rollup-plugin-node-resolve**
```javascript
  'use strict';

  Object.defineProperty(exports, '__esModule', { value: true });

  var isString_ts = require('./modules/is-string.ts');
 //                                                  see it here  ^^


  exports.isString = isString_ts;
  //# sourceMappingURL=index.js.map
```

**Comment by TrySound** [comment]
Does it happen for you with local modules? Why these files are not bundled?

**Comment by harrison-ifeanyichukwu** [comment]
@TrySound  the essence is not to bundle the files, but rather generate a clean cjs build, where all import statements are replaced with nodejs cjs buil. Thanks to the external array option that rollup accepts. Rollup is a great tool for this.  There is a great article regarding this here

https://medium.com/@kelin2025/so-you-wanna-use-es6-modules-714f48b3a953.

**Comment by harrison-ifeanyichukwu** [comment]
Also, like I pointed out, I was able to solve this issue just inside my node_modules rollup folder. I also pointed the file location where this is taking place [https://github.com/rollup/rollup/blob/master/src/Chunk.ts#L578](https://github.com/rollup/rollup/blob/master/src/Chunk.ts#L578)

**Comment by harrison-ifeanyichukwu** [comment]
it is just a matter of making sure that `relPath` in that line of code ends with `.js` file extension before it is assigned to the `id` property

`renderedDependency.id = relPath;`

. This is necessary because someone can be working with typescript, coffescript, etc, which will end up being transformed/transpiled to javascript

**this code solves it**
```javascript
let relPath = this.id ? normalize(relative(dirname(this.id), depId)) : depId;
if (!relPath.startsWith('../')) relPath = './' + relPath;

if (dep instanceof Chunk) renderedDependency.namedExportsMode = dep.exportMode !== 'default';

//--------------------------------------------------------------------------
const ext = path.extname(relPath);
relPath = relPath.slice(0, relPath.lastIndexOf(ext)) + '.js';
//---------------------------------------------------------------------------

renderedDependency.id = relPath;
```

**Comment by TrySound** [comment]
I don't quite get. So do you compile each file with rollup+ts? Why not just use ts? Rollup is about bundling, not converting esm to cjs.

**Comment by harrison-ifeanyichukwu** [comment]
I compile each file with rollup + babel typescript preset.

**Comment by harrison-ifeanyichukwu** [comment]
the only unique thing here is that I do not bundle, I only bundle the code when I need a browser iife version.

**Comment by harrison-ifeanyichukwu** [comment]
this is achievable because of the **external config option** that rollup allows. That is to say, when rollup finds an import statement, I tell rollup to consider the imported module as an external dependency. which makes rollup to replace the es6 import statement to node.js require statement.

The only fault here is that rollup ends up using the absolute path, which it then renormalizes back to relative path. but then, it has introduced the file extension

**Comment by TrySound** [comment]
Why not just use babel-cli? Rollup does nothing here except rendering specific format. Babel can do this just fine.

The article is weird. But probably the point is not bundling vendor dependencies to not duplicate them with other packages. Ignoring local modules is useless if you have just one entry point. Or do you provide path imports like `lodash/isString`?

**Comment by harrison-ifeanyichukwu** [comment]
babel-cli does this too, that very true. but rollup can too. I want to stick with rollup for all scenerios, both bundling and simply keeping my build intact, after rollup-plugin-babel is done processing the files through rollup

**Comment by harrison-ifeanyichukwu** [comment]
it is simple, rollup provided the external option, so why does it end up changing the external import statement if it is a local file?

**Comment by harrison-ifeanyichukwu** [comment]
It should not add any file extension to the path. just keep it as it is.

like
```javascript
import aModule from '../index';
```

should be turned to the code below if I tell rollup to consider '../index' as an external module. just like when I tell it that `import path from 'path'` is an  external module.
```javascript
var aModule = require('../index');
```

**Comment by harrison-ifeanyichukwu** [comment]
@TrySound I have been trying very well to explain what I mean here. Let me try again, and be more relaxed.

What I am saying is this.

First, the essence of rollup-plugin-babel is to integrate babel into the process, so that babel can transpile my es6, provide pollyfills for my build target if necessary, change import statements to require statements for node.js etc. good.

then babel-preset-typescript is another middleman that will stripout all typescript codes. good.

then rollup comes in. for rollup to work, we also tell babel to keep all our module import statements, that is, it should not replace imports with require statements.

now when rollup starts, it inspects our code for these import statements, and brings the codes into one piece, ensuring that there are no duplicates etc.

but, rollup also allows us to tell it to skip some imports, because they are external modules. such modules like path, fs, crypto, etc are node.js internals. We can also tell rollup to consider local files as externals too.

when this is done, rollup is smart enough to now replace the es6 import with the nodejs cjs require equivalent. this is super with modules like fs, crypto, etc.

but when it is a local file, rollup will then introduce the file extension even when it was not in the original. such that typescript files ends up having .ts in the require statement. That is bad.

**Comment by harrison-ifeanyichukwu** [comment]
@TrySound I hope the explanation above is clearer

**Comment by TrySound** [comment]
I understand the bug you are reporting and agree with it.

I don't get this
> now when rollup starts, it inspects our code for these import statements, and brings the codes into one piece, ensuring that there are no duplicates etc.

Do you have duplicated imports in each of your source files? You should fix it in source code and enforce with eslint or anything else.

In your pipeline rollup does nothing but rendering format which can be done by babel which you already use. So rollup just used for nothing in your case.

**Comment by harrison-ifeanyichukwu** [comment]
There at no duplicate imports. Except that rollup introduced file extension
in the final code. Which I have explained and pointed out where it is
happening. Inside the render method in Chunks class.

that's all I need you to fix. Thanks

**Comment by lukastaegert** [comment]
I am just on the go so I cannot check for sure but I think the issue is how you mark your imports as external. If you want to mark an import as external without changing it, you need to mark the import external as it is written, e.g via a function. My guess is that the imports without extensions that Typescript uses cannot be resolved directly by rollup but also do not match your pattern. So Rollup uses its plugins to resolve them to an external id. Only then does your array kick in, but now the files do have extensions. There is no rule in JavaScript that a file needs to have a .js extension or any extension at all which is why IMO it is wrong to guess an extension for Rollup and which is why you end up with .ts extensions. My suggestion is to use the function syntax of the external option

**Comment by harrison-ifeanyichukwu** [comment]
Wow, thanks lukas for this update. When I looked into the use of function,
rollup passes three arguments to the function,

1. the id, which I think it has already normalized like i pointed out

2. parentId

3. isResolved boolean value

Let me check with this, and update you. Thank you

**Comment by harrison-ifeanyichukwu** [comment]
@lukastaegert, that works. using a function solves this. it also makes rollup introduce the default interop function to stay save. Thank you. I will use functions onwards
